### PR TITLE
OSL Misc work

### PR DESCRIPTION
**Pull request by est77**
Added OSL's coordinate systems transforms from object, camera and world to NDC and raster spaces.

OSL's SurfaceShaders:
A surface shader that runs an OSL shader to post-process the shading results. Might be useful to do color corrections and other special effects.
The commit includes useful things, like marking OSL shader parameters as outputs and getting shader symbols to be able to fetch results from them.

### CODE REVIEW

File: src/appleseed/renderer/kernel/rendering/rendererservices.cpp [hunk 1]
```diff
@@ -347,6 +354,72 @@ bool RendererServices::transform_points(
...
+    // We only transform points for now.
+    if (vectype != OSL::TypeDesc::POINT)
+        return false;
+
+    if (to == g_NDC_ustr || to == g_raster_ustr)
+    {
+        if (from == g_world_ustr || from == g_common_ustr || from == g_shader_ustr)
+        {
+            // Copy the points to Pout.
+            for (size_t i = 0; i < npoints; ++i)
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/rendererservices.cpp]: Signed/unsigned mismatch here (i is `size_t`, npoints is `int`).

File: src/appleseed/renderer/kernel/shading/shadingpoint.h [hunk 2]
```diff
@@ -259,6 +259,7 @@ class ShadingPoint
     friend class CurveLeafVisitor;
     friend class Intersector;
     friend class OSLShaderGroupExec;
+    friend class RendererServices;
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/shadingpoint.h]: Order alphabetically.

File: src/appleseed/renderer/modeling/shadergroup/shadergroup.h [hunk 3]
```diff
@@ -134,6 +134,10 @@ class APPLESEED_DLLSYMBOL ShaderGroup
     // Return a reference-counted (but opaque) reference to the internal OSL shader group.
     OSL::ShaderGroupRef& shader_group_ref() const;
 
+    // OSL Symbols
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/shadergroup/shadergroup.h]: `// OSL symbols.`

File: src/appleseed/renderer/modeling/surfaceshader/oslsurfaceshader.cpp [hunk 4]
```diff
@@ -0,0 +1,246 @@
+
+//
+// This source file is part of appleseed.
+// Visit http://appleseedhq.net/ for additional information and resources.
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2015 Esteban Tovagliari, The appleseedhq Organization
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/surfaceshader/oslsurfaceshader.cpp]: Fix year in copyright notice (it probably should say 2016, or 2015-2016, until I update all notices for the year 2017).

File: src/appleseed/renderer/modeling/surfaceshader/oslsurfaceshader.cpp [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+        {
+            m_surface_shader = 0;
+            m_lighting_conditions = 0;
+            m_shader_group = 0;
+
+            SurfaceShader::on_frame_end(project, parent);
+        }
+
+        virtual void evaluate(
+            SamplingContext&        sampling_context,
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/surfaceshader/oslsurfaceshader.cpp]: Align all arguments vertically.

File: src/appleseed/renderer/modeling/surfaceshader/oslsurfaceshader.cpp [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+            const PixelContext&     pixel_context,
+            const ShadingContext&   shading_context,
+            const ShadingPoint&     shading_point,
+            ShadingResult&          shading_result) const APPLESEED_OVERRIDE
+        {
+            if (m_shader_group)
+            {
+                const OSL::ShaderSymbol *color_sym =
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/surfaceshader/oslsurfaceshader.cpp]: Star goes on the type's side.

**Comment by dictoon** [comment]
Pretty cool!

Can you provide an example of a postprocessing OSL surface shader?

**Comment by est77** [comment]
Fixed and rebased.
I don't have any example now of OSL surface shaders. I'll see if I can borrow some asset and make one.
I was more interested in merging the code that deals with OSL symbols and shader outputs.
We'll need it for light attenuation shaders, AOVs and possibly other uses.

**Comment by dictoon** [comment]
I suggest adding a few test scenes using very simple OSL surface shaders (e.g. a surface shader that does nothing, or just multiplies RGB by 2, or something like that) to make sure this code works and continues to do so in the future.

**Comment by est77** [comment]
Ok. I'll add some test scenes in a following PR.

**Pull request merged**